src/python/pants/goal/completion.py

```diff
@@ -0,0 +1,184 @@
+        show the available scoped options for the previous goal.
+        - `pants fmt -<tab>` -> `... --only ...`
+
+        :param options: The options object for the current Pants run.
+        :return: A list of completion options.
+        """
+        logger.debug(f"Completion passthrough options: {options._passthru}")
+        args = [arg for arg in options._passthru if arg != "pants"]
+        current_word = args.pop()
+        previous_goal = self._get_previous_goal(args)
+        logger.debug(f"Current word is '{current_word}', and previous goal is '{previous_goal}'")
+
+        all_goals = sorted([k for k, v in options.known_scope_to_info.items() if v.is_goal])
+
+        # If there is no previous goal, then we're at the top-level Pants command, so show all goals or global options
+        if not previous_goal:
+            if current_word.startswith("-"):
+                global_options = self._build_options_for_goal(options)
+                candidate_options = [o for o in global_options if o.startswith(current_word)]
+                return candidate_options
+
+            candidate_goals = [g for g in all_goals if g.startswith(current_word)]
+            return candidate_goals
+
+        # If there is already a previous goal and current_word starts with a hyphen, then show scoped options for that goal
+        if current_word.startswith("-"):
+            scoped_options = self._build_options_for_goal(options, previous_goal)
+            candidate_options = [o for o in scoped_options if o.startswith(current_word)]
+            return candidate_options
+
+        # If there is a previous goal and current_word does not start with a hyphen, then show remaining goals
+        # excluding the goals that are already in the command
+        candidate_goals = [
+            g for g in all_goals if g.startswith(current_word) and g not in options._passthru
+        ]
+        return candidate_goals
+
+    def _get_previous_goal(self, args: list[str]) -> str | None:
+        """Get the most recent goal in the command arguments, so options can be correctly applied.
+
+        A "goal" in the context of completions is simply an arg where the first character is alphanumeric.
+        This under-specifies the goal, because detecting whether an arg is an "actual" goal happens elsewhere.
+
+        :param args: The list of arguments to search for the previous goal.
+        :return: The previous goal, or None if there is no previous goal.
+        """
+        return next((arg for arg in reversed(args) if arg[:1].isalnum()), None)
+
+    def _build_options_for_goal(self, options: Options, goal: str = "") -> list[str]:
+        """Build a list of stringified options for the specified goal, prefixed by `--`.
+
+        :param options: The options object for the current Pants run.
+        :param goal: The goal to build options for. Defaults to "" for the global scope.
+        :return: A list of options for the specified goal.
+        """
```

Comment:
perhaps not for now, but I imagine we could look at the kind of option to determine if it may only be used once and already present, to exclude it from the list of completions. if that makes sense to do... on second thought, perhaps not worth the complexity.

Reply:
Oh yeah, that would be nice. But agreed on the complexity, it would be a lot messier than the goals check. 

Maybe https://github.com/pantsbuild/pants/issues/20676